node/src/ledger/ledger_service.rs

```diff
@@ -324,6 +334,123 @@ impl<T: LedgerService> TransitionFrontierService for T {
         Ok(())
     }
 
+    fn push_snarked_ledger(
+        &mut self,
+        protocol_states: &BTreeMap<StateHash, MinaStateProtocolStateValueStableV2>,
+        old_root: &ArcBlockWithHash,
+        new_root: &ArcBlockWithHash,
+    ) -> Result<(), String> {
+        // Steps 4-7 from https://github.com/openmina/mina/blob/bc812dc9b90e05898c0c36ac76ba51ccf6cac137/src/lib/transition_frontier/full_frontier/full_frontier.ml#L354-L392
+        let ctx = self.ctx_mut();
+        let constraint_constants = &CONSTRAINT_CONSTANTS;
+
+        // Step 4: create a new temporary mask `mt` with `s` as it's parent
+        let root_snarked_ledger = ctx
+            .snarked_ledgers
+            .remove(old_root.snarked_ledger_hash())
+            .ok_or_else(|| format!(
+                "push_snarked_ledger: could not find old root snarked ledger: {} (in_staged: {}, in_sync_staged: {})",
+                old_root.snarked_ledger_hash().to_string(),
+                ctx.staged_ledgers.keys().any(|h| h == old_root.snarked_ledger_hash()),
+                ctx.sync.staged_ledgers.keys().any(|h| h == old_root.snarked_ledger_hash())
+            ))?;
+        let mut mt = root_snarked_ledger.make_child();
```

Comment:
Do we create this tmp mask so that eventually in case of error, we discard it and reinsert the old snarked ledger?

Reply:
@binier yes, you want to avoid mutating the original ledger until you are sure there will be no failures. Reinsertion would not be needed because it was not removed (in the same way the new one is not inserted until it is ready).

These failures should not happen in practice tho unless the implementation is buggy (but the way to figure out if it is buggy or not is to have these checks). All the transactions that are re-applied here have been applied before when they came in a block.

Btw note that right now the node cannot recover from this, in the caller there is a `panic!` if this `push_snarked_ledger` call fails with an error.